server/base/src/main/java/org/apache/accumulo/server/fs/PreferredVolumeChooser.java

```diff
@@ -21,84 +21,90 @@
-  // TODO ACCUMULO-3417 replace this with the ability to retrieve by String key.
-  private static final Predicate<String> PREFERRED_VOLUMES_FILTER = key -> PREFERRED_VOLUMES_CUSTOM_KEY.equals(key);
+  public static final String PREFERRED_VOLUMES_CUSTOM_KEY = Property.TABLE_ARBITRARY_PROP_PREFIX + "preferredVolumes";
 
   @SuppressWarnings("unchecked")
   private final Map<String,Set<String>> parsedPreferredVolumes = Collections.synchronizedMap(new LRUMap(1000));
   // TODO has to be lazily initialized currently because of the reliance on HdfsZooInstance. see ACCUMULO-3411
   private volatile ServerConfigurationFactory serverConfs;
 
   @Override
-  public String choose(VolumeChooserEnvironment env, String[] options) {
-    if (!env.hasTableId())
+  public String choose(VolumeChooserEnvironment env, String[] options) throws AccumuloException {
+    if (!env.hasTableId() && !env.hasScope()) {
+      // this should only happen during initialize
+      log.warn("No table id or scope, so it's not possible to determine preferred volumes.  Using all volumes.");
       return super.choose(env, options);
+    }
+    ServerConfigurationFactory localConf = loadConf();
 
-    // Get the current table's properties, and find the preferred volumes property
-    // This local variable is an intentional component of the single-check idiom.
-    ServerConfigurationFactory localConf = serverConfs;
-    if (localConf == null) {
-      // If we're under contention when first getting here we'll throw away some initializations.
-      localConf = new ServerConfigurationFactory(HdfsZooInstance.getInstance());
-      serverConfs = localConf;
+    String systemPreferredVolumes = localConf.getConfiguration().get(PREFERRED_VOLUMES_CUSTOM_KEY);
+    if (null == systemPreferredVolumes || systemPreferredVolumes.isEmpty()) {
+      String logMessage = "Default preferred volumes are missing.";
+      log.debug(logMessage);
+      throw new AccumuloException(logMessage);
     }
-    TableConfiguration tableConf = localConf.getTableConfiguration(env.getTableId());
-    final Map<String,String> props = new HashMap<>();
-    tableConf.getProperties(props, PREFERRED_VOLUMES_FILTER);
-    if (props.isEmpty()) {
-      log.warn("No preferred volumes specified. Defaulting to randomly choosing from instance volumes");
-      return super.choose(env, options);
+
+    String volumes = null;
+    if (env.hasTableId()) {
+      volumes = localConf.getTableConfiguration(env.getTableId()).get(PREFERRED_VOLUMES_CUSTOM_KEY);
+    } else if (env.hasScope()) {
+      volumes = localConf.getConfiguration().get(Property.GENERAL_ARBITRARY_PROP_PREFIX.getKey() + env.getScope() + ".preferredVolumes");
+    }
+
+    // if there was an empty or missing property, use the system-wide volumes
+    if (null == volumes || volumes.isEmpty()) {
+      if (env.hasTableId()) {
+        log.warn("Missing property for TableID " + env.getTableId() + " but it should have picked up default volumes.");
```

Comment:
It's probably a warning, because the current behavior falls back, but I think we'd want to change this so it's a hard fail. It seems it would only happen when there's a misconfiguration (or... without atomic config updates, if one property propagates via ZK faster than the corresponding properties).